Wrappers/Python/test/test_out_in_place.py

```diff
@@ -0,0 +1,310 @@
+            (MixedL21Norm(), bg),
+            (SmoothMixedL21Norm(epsilon=0.3), bg),
+            (MixedL11Norm(), bg)
+
+        ]
+
+        np.random.seed(5)
+        self.data_arrays=[np.random.normal(0,1, (10,10)).astype(np.float32),  np.array(range(0,65500, 655), dtype=np.uint16).reshape((10,10)), np.random.uniform(-0.1,1,(10,10)).astype(np.float32)]
+
+    def get_result(self, function, method, x, *args):
+        try:
+            input=x.copy() #To check that it isn't changed after function calls 
+            if method == 'proximal':
+                out= function.proximal(x, *args)
+            elif method == 'proximal_conjugate':
+                out= function.proximal_conjugate(x, *args)
+            elif method == 'gradient':
+                out= function.gradient(x, *args)
+            self.assertDataArraysInContainerAllClose(input, x, rtol=1e-5, msg= "In case func."+method+'(data, *args) where func is  ' + function.__class__.__name__+ 'the input data has been incorrectly affected by the calculation. ')
+            return out
+        except NotImplementedError:
+            return None
+        
+    def in_place_test(self,desired_result, function, method,   x, *args, ): 
+            out3 = x.copy()
+            try:
+                if method == 'proximal':
+                    function.proximal(out3, *args, out=out3)
+                elif method == 'proximal_conjugate':
+                    function.proximal_conjugate(out3, *args, out=out3)
+                elif method == 'gradient':
+                    function.gradient(out3, *args, out=out3)
+                self.assertDataArraysInContainerAllClose(desired_result, out3, rtol=1e-5, msg= "In place calculation failed for func."+method+'(data, *args, out=data) where func is  ' + function.__class__.__name__+ '. ')
+      
+            except (InPlaceError, NotImplementedError):
+                pass
+ 
+        
+    def out_test(self, desired_result, function, method,  x, *args, ):
+        input = x.copy()
+        out2=0*(x.copy())
+        try:
+            if method == 'proximal':
+                function.proximal(input, *args, out=out2)
+            elif method == 'proximal_conjugate':
+                function.proximal_conjugate(input, *args, out=out2)
+            elif method == 'gradient':
+                function.gradient(input, *args, out=out2)
+            self.assertDataArraysInContainerAllClose(desired_result, out2, rtol=1e-5, msg= "Calculation failed using `out` in func."+method+'(x, *args, out=data) where func is  ' + function.__class__.__name__+ '. ')
+            self.assertDataArraysInContainerAllClose(input, x,  rtol=1e-5, msg= "In case func."+method+'(data, *args, out=out) where func is  ' + function.__class__.__name__+ 'the input data has been incorrectly affected by the calculation. ')
+      
+        except (InPlaceError, NotImplementedError):
+            pass
+            
+            
+   
+    def test_proximal_conjugate_out(self):
```

Comment:
The loop looks really useful! Just wondering why you choose to test the proximal, proximal conjugate and gradient in separate loops?

Reply:
Good question. I think it is just so the user can more clearly see where things will go wrong. They will see an error in the test_proximal_conjugate_out and know it is an issue with the proximal conjugate. It helps with debugging.